array.c

```diff
@@ -4692,198 +4692,251 @@ rb_ary_clear(VALUE ary)
+ *      a.fill('-', 1, 2)   # => ["a", "-", "-", "d"]
  *
- *  With arguments +obj+ and Range +range+, and no block given,
- *  replaces elements based on the given range.
+ *  - If +start+ is negative, counts from the end:
  *
- *  If the range is positive and ascending (<tt>0 < range.begin <= range.end</tt>),
- *  replaces elements from <tt>range.begin</tt> to <tt>range.end</tt>:
+ *      a = ('a'..'d').to_a # => ["a", "b", "c", "d"]
+ *      a.fill('-', -3, 2)  # => ["a", "-", "-", "d"]
  *
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, (1..1)) # => ["a", :X, "c", "d"]
+ *  - If +start+ is large (<tt>start >= array.size</tt>), extends +self+ with +nil+:
  *
- *  If <tt>range.first</tt> is negative, replaces no elements:
+ *      a = ('a'..'d').to_a # => ["a", "b", "c", "d"]
+ *      a.fill('-', 6, 2)   # => ["a", "b", "c", "d", nil, nil, "-", "-"]
+ *      # Extends even if length is zero.
+ *      a = ('a'..'d').to_a # => ["a", "b", "c", "d"]
+ *      a.fill('-', 6, 0)   # => ["a", "b", "c", "d", nil, nil]
  *
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, (-1..1)) # => ["a", "b", "c", "d"]
+ *  - If +length+ is zero or negative, replaces no elements:
  *
- *  If <tt>range.last</tt> is negative, counts from the end:
+ *      a = ('a'..'d').to_a # => ["a", "b", "c", "d"]
+ *      a.fill('-', 2, 0)   # => ["a", "b", "c", "d"]
+ *      a.fill('-', 2, -1)  # => ["a", "b", "c", "d"]
  *
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, (0..-2)) # => [:X, :X, :X, "d"]
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, (1..-2)) # => ["a", :X, :X, "d"]
+ *  With argument +object+ and Range argument +range+,
+ *  replaces elements based on the given range:
  *
- *  If <tt>range.last</tt> and <tt>range.last</tt> are both negative,
- *  both count from the end of the array:
+ *  - If the range is positive and ascending (<tt>0 < range.begin <= range.end</tt>),
+ *    replaces elements from <tt>range.begin</tt> to <tt>range.end</tt>:
  *
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, (-1..-1)) # => ["a", "b", "c", :X]
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, (-2..-2)) # => ["a", "b", :X, "d"]
+ *      a = ('a'..'d').to_a # => ["a", "b", "c", "d"]
+ *      a.fill('-', (1..2)) # => ["a", "-", "-", "d"]
  *
- *  With no arguments and a block given, calls the block with each index;
+ *  - If <tt>range.first</tt> is negative, replaces no elements:
+ *
+ *      a = ('a'..'d').to_a  # => ["a", "b", "c", "d"]
+ *      a.fill('-', (-1..2)) # => ["a", "b", "c", "d"]
+ *
+ *  - If <tt>range.last</tt> is negative, counts from the end:
+ *
+ *      a = ('a'..'d').to_a  # => ["a", "b", "c", "d"]
+ *      a.fill('-', (1..-2)) # => ["a", "-", "-", "d"]
+ *
+ *  - If <tt>range.last</tt> and <tt>range.last</tt> are both negative,
+ *    both count from the end of the array:
+ *
+ *      a = ('a'..'d').to_a   # => ["a", "b", "c", "d"]
+ *      a.fill('-', (-3..-2)) # => ["a", "-", "-", "d"]
+ *
+ *  <b>With a Block Given</b>
```

Comment:
It seems like with a block works the same as without a block except it calls the block for each replacement instead of using the predefined `object`? In that case, I don't think we need to duplicate the whole documentation for this.

Reply:
Agreed;  with-block part should be much briefer.